source/mqtt_subscription_set.c

```diff
@@ -0,0 +1,431 @@
+    aws_hash_table_foreach(&node->children, s_subscription_set_node_destroy_hash_foreach_wrap, NULL);
+    aws_hash_table_clean_up(&node->children);
+
+    if (node->on_cleanup && node->callback_user_data) {
+        node->on_cleanup(node->callback_user_data);
+    }
+
+    aws_byte_buf_clean_up(&node->topic_segment);
+
+    aws_mem_release(node->allocator, node);
+}
+
+static void s_aws_mqtt_subscription_set_node_destroy_tree(struct aws_mqtt_subscription_set_topic_tree_node *tree) {
+    if (tree == NULL) {
+        return;
+    }
+
+    if (tree->parent != NULL) {
+        aws_hash_table_remove(&tree->parent->children, &tree->topic_segment, NULL, NULL);
+    }
+
+    s_aws_mqtt_subscription_set_node_destroy_node(tree);
+}
+
+static int s_subscription_set_node_destroy_hash_foreach_wrap(void *context, struct aws_hash_element *elem) {
+    (void)context;
+
+    s_aws_mqtt_subscription_set_node_destroy_node(elem->value);
+
+    return AWS_COMMON_HASH_TABLE_ITER_CONTINUE | AWS_COMMON_HASH_TABLE_ITER_DELETE;
+}
+
+void aws_mqtt_subscription_set_destroy(struct aws_mqtt_subscription_set *subscription_set) {
+    if (subscription_set == NULL) {
+        return;
+    }
+
+    s_aws_mqtt_subscription_set_node_destroy_tree(subscription_set->root);
+
+    aws_hash_table_clean_up(&subscription_set->subscriptions);
+
+    aws_mem_release(subscription_set->allocator, subscription_set);
+}
+
+static struct aws_mqtt_subscription_set_topic_tree_node *s_aws_mqtt_subscription_set_get_existing_subscription_node(
+    struct aws_mqtt_subscription_set *subscription_set,
+    struct aws_byte_cursor topic_filter) {
+
+    struct aws_mqtt_subscription_set_topic_tree_node *current_node = subscription_set->root;
+
+    struct aws_byte_cursor topic_segment;
+    AWS_ZERO_STRUCT(topic_segment);
+    while (aws_byte_cursor_next_split(&topic_filter, '/', &topic_segment)) {
+        struct aws_hash_element *hash_element = NULL;
+        aws_hash_table_find(&current_node->children, &topic_segment, &hash_element);
+
+        if (hash_element == NULL) {
+            return NULL;
+        } else {
+            current_node = hash_element->value;
+        }
+    }
+
+    if (!current_node->is_subscription) {
+        return NULL;
+    }
+
+    return current_node;
+}
+
+bool aws_mqtt_subscription_set_is_subscribed(
+    struct aws_mqtt_subscription_set *subscription_set,
+    struct aws_byte_cursor topic_filter) {
+
+    struct aws_hash_element *element = NULL;
+    aws_hash_table_find(&subscription_set->subscriptions, &topic_filter, &element);
+
+    return element && (element->value != NULL);
+}
+
+bool aws_mqtt_subscription_set_is_in_topic_tree(
+    struct aws_mqtt_subscription_set *subscription_set,
+    struct aws_byte_cursor topic_filter) {
+    struct aws_mqtt_subscription_set_topic_tree_node *existing_node =
+        s_aws_mqtt_subscription_set_get_existing_subscription_node(subscription_set, topic_filter);
+
+    return existing_node != NULL;
+}
+
+/*
```

Comment:
Trivial - should this comment/description be in the header instead of here?

Reply:
The comment as a whole is pretty tied to the internal function but an additional blurb about how the ref counting works in the structure declaration would be useful.